skyportal/handlers/api/allocation.py

```diff
@@ -222,3 +228,139 @@ def delete(self, allocation_id):
         self.verify_and_commit()
         self.push_all(action='skyportal/REFRESH_ALLOCATIONS')
         return self.success()
+
+
+class AllocationAnalysisHandler(BaseHandler):
+    @auth_or_token
+    async def get(self, instrument_id):
+        """
+        ---
+        tags:
+          - allocations
+        description: Produce a report on allocations for an instrument
+        parameters:
+          - in: path
+            name: instrument_id
+            required: true
+            schema:
+              type: integer
+          - in: query
+            name: output_format
+            nullable: true
+            schema:
+              type: string
+            description: |
+              Output format for analysis. Can be png or pdf
+        responses:
+          200:
+            content:
+              application/json:
+                schema: Success
+          400:
+            content:
+              application/json:
+                schema: Error
+        """
+
+        output_format = self.get_query_argument('output_format', 'pdf')
+        if output_format not in ["pdf", "png"]:
+            return self.error('output_format must be png or pdf')
+
+        # get owned allocations
+        allocations = Allocation.query_records_accessible_by(self.current_user)
+        allocations = allocations.filter(
+            Allocation.instrument_id == instrument_id
+        ).all()
+
+        if len(allocations) == 0:
+            return self.error('Need at least one allocation for analysis')
+
+        instrument = allocations[0].instrument
+
+        labels = [a.proposal_id for a in allocations]
+        values = [a.hours_allocated for a in allocations]
+
+        def make_autopct_hours(values):
+            def my_autopct(pct):
+                total = sum(values)
+                if np.isnan(pct):
+                    val = 0
+                else:
+                    val = int(round(pct * total / 100.0))
+                return f'{pct:.0f}%  ({val:d} Hours)'
+
+            return my_autopct
+
+        def make_autopct_requests(values):
+            def my_autopct(pct):
+                total = sum(values)
+                if np.isnan(pct):
+                    val = 0
+                else:
+                    val = int(round(pct * total / 100.0))
+                return f'{pct:.0f}%  ({val:d} Requests)'
+
+            return my_autopct
+
+        def make_autopct_observations(values):
+            def my_autopct(pct):
+                total = sum(values)
+                if np.isnan(pct):
+                    val = 0
+                else:
+                    val = int(round(pct * total / 100.0))
+                return f'{pct:.0f}%  ({val:d} Observations)'
+
+            return my_autopct
```

Comment:
Is there an easy way to not duplicate all these functions?

Reply:
Done.